contracts/0.8.25/vaults/VaultHub.sol

```diff
@@ -451,162 +479,56 @@ contract VaultHub is PausableUntilWithRoles {
-                    // uint256 lockedEther = newMintedStETH * TOTAL_BASIS_POINTS / (TOTAL_BASIS_POINTS - socket.reserveRatioBP);
-                    (newMintedShares * _postInternalEther * TOTAL_BASIS_POINTS)
-                        / (_postInternalShares * (TOTAL_BASIS_POINTS - socket.reserveRatioBP)),
-                    CONNECT_DEPOSIT
-                );
-            }
-        }
-    }
+        bytes32 root = $.vaultsDataTreeRoot;
+        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(_vault, _valuation, _inOutDelta, _feeSharesCharged, _sharesMinted))));
+        if (!MerkleProof.verify(_proof, root, leaf)) revert InvalidProof();
 
-    /// @notice calculates the amount of shares to mint as treasury fees for the vault
-    /// @param _reportValuation the valuation of the vault from the report
-    /// @param socket the socket of the vault
-    /// @param _preTotalShares the total shares of the Lido protocol before the report
-    /// @param _preTotalPooledEther the total pooled ether of the Lido protocol before the report
-    /// @param _postInternalShares the internal shares of the Lido protocol after the report
-    /// @param _postInternalEther the internal ether of the Lido protocol after the report
-    /// @param _sharesToMintAsLidoCoreFees the amount of shares that is minted as the total Lido core fees (treasury and NO)
-    /// @return treasuryFeeShares the amount of shares to mint as treasury fees or 0 if _sharesToMintAsLidoCoreFees is 0
-    function calculateVaultTreasuryFees(
-        uint256 _reportValuation,
-        VaultSocket memory socket,
-
-        uint256 _preTotalShares,
-        uint256 _preTotalPooledEther,
-        uint256 _postInternalShares,
-        uint256 _postInternalEther,
-        uint256 _sharesToMintAsLidoCoreFees
-    ) public pure returns (uint256 treasuryFeeShares) {
-        // if lido doesn't charge the protocol fees, vaults don't charge either
-        if (_sharesToMintAsLidoCoreFees == 0) {
-            return 0;
+        VaultSocket storage socket = $.sockets[vaultIndex];
+        // NB: charged fees can only cumulatively increase with time
+        if (_feeSharesCharged  < socket.feeSharesCharged) {
+            revert InvalidFees(_vault, _feeSharesCharged, socket.feeSharesCharged);
         }
+        socket.sharesMinted += uint96(_feeSharesCharged - socket.feeSharesCharged);
+        socket.feeSharesCharged = uint96(_feeSharesCharged);
 
-        uint256 mintableRatio = (TOTAL_BASIS_POINTS - socket.reserveRatioBP);
-
-        uint256 chargeableValuation = Math256.min(
-            // we are charging fees over the mintable part of the vault's valuation
-            _reportValuation * mintableRatio / TOTAL_BASIS_POINTS,
-            // capped by the vault's shareLimit
-            socket.shareLimit * _postInternalEther / _postInternalShares
+        uint256 newMintedShares = Math256.max(socket.sharesMinted, _sharesMinted);
+        uint256 lockedEther = Math256.max(
```

Comment:
```suggestion
        // locked ether can only be increased asynchronously once the oracle settled the new floor value 
        // as of reference slot to prevent slashing upsides in between the report gathering and delivering
        uint256 lockedEther = Math256.max(
```